### ISSUE TITLE
Add generic Network directives for network configuration

### ISSUE THREAD

**Issue by rpigott** [systemd/systemd#32273]
### Component

systemd-networkd

### Is your feature request related to a problem? Please describe

If I want to not use the DNS offered by the local network configuration, currently there are 3 options to set: DHCPv4.UseDNS, DHCPv6.UseDNS, IPv6AcceptRA.UseDNS. Worse, if more configuration protocols were to be introduced, perhaps DNR, or IPCP (#481), or some future network config protocol, it would become ineffective, because any new configuration protocol will circumvent the user's choice not to use some local network configuration as a matter of policy, even if sd-networkd is technically working correctly as directed. The same is true for other kinds of network configuration, domains, NTP servers, etc. 

For novice users, it might not obvious that there are three settings to consider in order to effect this decision as evidenced by various distro forum support posts. Generally speaking, most of the time we won't care exactly how the local network informed us of some configuration when making a decision to utilize it.

### Describe the solution you'd like

There really should be one setting that controls all three (or more) values. This will enable the user to make the decision to use local network configurations as a matter of policy. I'm proposing new .network configuration directives:

```
[Network]
UseDNS=
UseDomains=
[...]
```

etc. that are similar to existing directives but under the Network section. These control all 3 (or more) protocol specific values unless overridden by a more specific directive.

### Describe alternatives you've considered

The status quo. We should also consider whether or not more protocol specific and generic options are necessary in the manager configuration such as in #32194.

### The systemd version you checked that didn't have the feature you are asking for

255.4

**Comment by yuwata** [comment]
I openend #32276 for UseDomains=. If you like it, I will add similar settings later.

**Comment by evverx** [comment]
FWIW I'm still not sure it should be possible to set, say, `UseDomains=` in `networkd.conf`. It defaults to `no` for a reason and it's kind of weird to make it easier to set to `yes` globally. Other than that the global config can't be reloaded with `networkctl reload` so it makes it less flexible than the other configs. Then again as long as this thing is off by default I don't care much.

**Comment by yuwata** [comment]
@evverx We should definitely support reloading `networkd.conf` by `networkctl reload`. But let's do that later. If you'd like, feel free to open a new RFE about that.

**Comment by evverx** [comment]
I explicitly disable things in `*.network` files even when they default to `no` so I don't think I have ever needed to set those things globally and reload the global config. I guess I just don't understand what the use case is. It isn't explained in https://github.com/systemd/systemd/pull/32194 either.

**Comment by yuwata** [comment]
> I explicitly disable things in `*.network` files even when they default to `no` so I don't think I have ever needed to set those things globally and reload the global config. I guess I just don't understand what the use case is. It isn't explained in #32194 either.

That's simply like DefaultXYZ= in systemd-system.conf(5).

**Comment by evverx** [comment]
I get that and I can see why it makes sense to set those settings globally. I don't understand why it would be necessary to set `UseDomains=` (and only `UseDomains=` and only for DHCPv4/DHCPv6) globally. It looks weirdly specific to me.

**Comment by yuwata** [comment]
Not only for UseDomains=, but it may be useful to override default settings, like `DHCP=`, `IPv6AcceptRA=`, `UseNTP=`,...

**Comment by evverx** [comment]
Dunno. I think it would be great to have generic `UseDomains=`, `UseDNS=` and so on and be able to set them in `*.network` files (because for example when I point the fuzzer at networkd I'm interested in covering as much code as possible and generally don't care whether it crashes via DHCPv4, DHCPv6, router advertisements or whatever) but I'm not sure how useful it is to set those defaults globally. Then again I don't feel strongly about this.

**Comment by rpigott** [comment]
I agree with @evverx. IMO the purview of sd-networkd is a little too large for most settings to have sensible defaults set in the manager. With e.g. DHCP=yes in the manager we would start a DHCP client by default on every managed link including nonsense cases like wireguard or sit tunnels. These settings may or may not be a marginal simplification for some deployments, but I think they are an enormous footgun for regular users.

I wonder if it would be possible to introduce a more comprehensive drop-in mechanism for .network files to satisfy this demand. In the same way that systemd units support top-level drop-ins like `service.d` to apply to all service units, it might be handy to have e.g. ether.d/*.network be merged with all network files that apply to networks which matched Type=ether. That's obviously a bit more complex, but a UseDomains that applies to all Type=ether networks sounds much more reasonable to me.

I would actually like to see even more generalized network settings, too. In the same way that we have ProtectSystem= directive to roll-up disparate sandboxing settings, we could have some kind of "trusted"/"untrusted" presets for network files as well, that would enable usage of various network configurations and privacy settings. This would mimic how Windows distinguishes between "Private" and "Public" network profiles.

**Comment by kanashimia** [comment]
I've just encountered the same problem while helping a guy, he set UseDNS=false in DHCPv4, but forgot about DHCPv6, and then we both forgot about IPv6AcceptRA, was pretty annoying.

Although not as important, setting default value of RouteMetric in a link would also be nice, for example for a typical ether / wlan setup.
